cypress/support/auth-provider-commands/auth0.ts

```diff
@@ -1,53 +1,37 @@
 // @ts-check
 ///<reference path="../../global.d.ts" />
 
-import * as jwt from "jsonwebtoken";
-
-Cypress.Commands.add("loginByAuth0Api", (username: string, password?: string) => {
+// Note: this function leaves you on a blank page, so you must call cy.visit()
+// afterwards, before continuing with your test.
+Cypress.Commands.add("loginByAuth0Api", (username: string, password: string) => {
   const log = Cypress.log({
     displayName: "AUTH0 LOGIN",
     message: [`🔐 Authenticating | ${username}`],
     // @ts-ignore
     autoEnd: false,
   });
-
-  const client_id = Cypress.env("auth0_client_id");
-  const client_secret = Cypress.env("auth0_client_secret");
-  const audience = Cypress.env("auth0_audience");
-  const scope = Cypress.env("auth0_scope");
-
   log.snapshot("before");
 
-  cy.request({
-    method: "POST",
-    url: `https://${Cypress.env("auth0_domain")}/oauth/token`,
-    body: {
-      grant_type: "password",
-      username,
-      password,
-      audience,
-      scope,
-      client_id,
-      client_secret,
-    },
-  }).then(({ body }) => {
-    const user: any = jwt.decode(body.id_token);
-
-    const userItem = {
-      token: body.access_token,
-      user: {
-        sub: user.sub,
-        nickname: user.nickname,
-        picture: user.name,
-        email: user.email,
-      },
-    };
-
-    window.localStorage.setItem("auth0Cypress", JSON.stringify(userItem));
-
-    log.snapshot("after");
-    log.end();
+  const args = { username, password };
+  cy.session(args, () => {
+    // App landing page redirects to Auth0.
+    cy.visit("/");
+
+    // Login on Auth0.
+    cy.origin("auth0.com", { args }, ({ username, password }) => {
+      cy.get("input#username").type(username);
+      cy.get("input#password").type(password);
+      cy.contains("button[value=default]", "Continue").click();
+    });
+    // Auth0 redirects back to RWA.
+
+    // Wait for RWA to save auth token to localstorage before saving session.
+    cy.url().should((url) => {
+      expect(url).to.contain(Cypress.config("baseUrl"));
+      expect(localStorage.getItem("authAccessToken")).to.exist;
+    });
   });
```

Comment:
This isn't quite what I'd expect. I would have thought this should be testing that RWA page load 

and the validate function t be verifying the `authAccessToken `

Reply:
The purpose of this code is to keep the session from being saved until the RWA has had time to write its auth token to localstorage. Without it we end up with a race condition, sometimes the RWA saves the token in time, sometimes it doesn't. Putting the code in the `validate` callback doesn't work because validate runs after the session has already saved.

Reply:
I think this is fine, you have to ensure the session is setup correctly before storing it.

You may consider also adding a validate function but it's purpose is other than what you're doing here. It's to ensure that the cached session is still valid.